File: ai-service/app/models/penalty.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Integer, Float, DateTime, ForeignKey, Text
from app.database import Base

class PenaltyRecord(Base):
    __tablename__ = "penalty_records"

    id = Column(String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    achievement_id = Column(String(50), ForeignKey("achievements.id"), nullable=False, index=True)
    team_id = Column(String(50), ForeignKey("teams.id"), nullable=False, index=True)
    
    original_points = Column(Integer, nullable=False)
    penalty_rate = Column(Float, nullable=False, default=0.90)  # Official 90% penalty rate
    penalty_points = Column(Integer, nullable=False)  # original_points * 0.90
    final_team_points = Column(Integer, nullable=False)  # original_points - penalty_points
    
    reason = Column(Text, nullable=False)
    verified_by = Column(String(50), ForeignKey("users.id"), nullable=False)
    applied_at = Column(DateTime, default=datetime.utcnow, index=True)

File: backend/src/models/penalty.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct PenaltyRecord {
    pub id: String,
    pub achievement_id: String,
    pub team_id: String,
    pub original_points: i32,
    pub penalty_rate: f64,
    pub penalty_points: i32,
    pub final_team_points: i32,
    pub reason: String,
    pub verified_by: String,
    pub applied_at: Option<NaiveDateTime>,
}
